dependency version fix: datasets

robinnarsinghranabhat opened the pull request:
By Default, datasets 4.0.0 is installed, which won't work. 

https://github.com/huggingface/datasets/issues/7693#issuecomment-3103380232

lhoestq commented:
I tried to convert the dataset to Parquet to make it work with `datasets>=4.0` but got this error for the "pooled" subset

```
Traceback (most recent call last):
  File "/Users/quentinlhoest/.pyenv/versions/3.11.6/envs/test/lib/python3.11/site-packages/datasets/builder.py", line 1608, in _prepare_split_single
    for key, record in generator:
  File "/Users/quentinlhoest/.cache/huggingface/modules/datasets_modules/datasets/colbertv2--lotte_passages/eca26a86e3b8ccbcf72fd2edc5c6555155f49b23226cca0d171f09c26066b901/lotte_passages.py", line 174, in _generate_examples
    data = json.loads(row)
           ^^^^^^^^^^^^^^^
  File "/Users/quentinlhoest/.pyenv/versions/3.11.6/lib/python3.11/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/quentinlhoest/.pyenv/versions/3.11.6/lib/python3.11/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/quentinlhoest/.pyenv/versions/3.11.6/lib/python3.11/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
               ^^^^^^^^^^^^^^^^^^^^^^
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 2562 (char 2561)
```

To reproduce:

```
uv run --with "datasets==3.6.0" datasets-cli convert_to_parquet colbertv2/lotte_passages --trust_remote_code
```

it looks like the dataset has amore general issue

vishalbakshi commented:
@robinnarsinghranabhat Thanks for flagging this. What's the use case where you're encountering an error? Can you provide some code for reproducibility?

robinnarsinghranabhat commented:
@vishalbakshi I built the conda-gpu environment from `conda env create -f conda_env.yml`. 

This by default didn't work. I used code you had used to test it  : 
https://github.com/stanford-futuredata/ColBERT/issues/371#issuecomment-2657380132


Furthermore, I was getting compilations issues. 
Overall, this worked for me : 
`https://github.com/stanford-futuredata/ColBERT/issues/371#issuecomment-3251906773`

vishalbakshi commented:
@robinnarsinghranabhat Gotcha, thanks for that information related to `conda_env.yml`. However, I'm still not clear what error you ran into for `datasets` which would require it to be pinned to `3.6.0`. Can you share what `datasets`-related error you got, for reproducibility?

robinnarsinghranabhat commented:
@vishalbakshi no need to be 3.6.0 exactly.   We just got to be wary, with the latest datasets version 4.0.0, it won't work, because of modern type hint syntax introduced in python 3.9+. Just updated `setup.py`.

I suppose this PR can be considered irrelavant, because I can't replicate, how datasets 4.0.0 got installed, the first time I setup colbert. I had to test with lots of install and uninstalls, and finally this [installation procedure](https://github.com/stanford-futuredata/ColBERT/issues/356#issuecomment-3263277117)  worked.  

For the time being, pip 21.0 will install datasets 3.1.0, and it's fine as well.


For example, following code  throws error with datasets 4.0.0 :
```
import colbert
from colbert import Indexer, Searcher
from colbert.infra import Run, RunConfig, ColBERTConfig
from colbert.data import Queries, Collection
from datasets import load_dataset

def main():
    nbits = 2  
    ndocs = 100
    dataset_name = "Genomics"
    index_name = f'{dataset_name}.{nbits}bits'

    passages = load_dataset("UKPLab/dapr", f"{dataset_name}-corpus", split="test")

    checkpoint = 'colbert-ir/colbertv2.0'

    with Run().context(RunConfig(nranks=1, experiment='notebook')):
        config = ColBERTConfig(doc_maxlen=10, nbits=nbits, kmeans_niters=2)
        indexer = Indexer(checkpoint=checkpoint, config=config)
        indexer.index(name=index_name, collection=passages[:100]["text"], overwrite=True)

if __name__ == '__main__':
    main()
```

Error : 
```
Traceback (most recent call last):
  File "test.py", line 5, in <module>
    from datasets import load_dataset
  File "miniforge3/envs/colbert/lib/python3.8/site-packages/datasets/__init__.py", line 17, in <module>
    from .arrow_dataset import Column, Dataset
  File "miniforge3/envs/colbert/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 75, in <module>
    from .arrow_reader import ArrowReader
  File "miniforge3/envs/colbert/lib/python3.8/site-packages/datasets/arrow_reader.py", line 30, in <module>
    from .download.download_config import DownloadConfig  # noqa: F401
  File "miniforge3/envs/colbert/lib/python3.8/site-packages/datasets/download/__init__.py", line 8, in <module>
    from .download_config import DownloadConfig
  File "miniforge3/envs/colbert/lib/python3.8/site-packages/datasets/download/download_config.py", line 10, in <module>
    class DownloadConfig:
  File "miniforge3/envs/colbert/lib/python3.8/site-packages/datasets/download/download_config.py", line 68, in DownloadConfig
    storage_options: dict[str, Any] = field(default_factory=dict)
TypeError: 'type' object is not subscriptable
```
